pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -212,20 +242,215 @@ func (c *MCDefaultRouteController) processNextWorkItem() bool {
+//  1. If the key is a Gateway dummy key, it means that we should check current Node and initialize/remove corresponding
+//     WireGuard interface on the Gateway Node.
+//  2. If the key is not a Gateway dummy key, it means that a new ClusterInfoImport event comes. The controller should
+//     update WireGuard peers on the Gateway Node.
+//
+// Note: MCDefaultRouteController runs only one worker to process Gateway and ClusterInfoImport. So we do not need
+// any synchronization mechanism.
+func (c *MCDefaultRouteController) reconcileWireGuard(key string) error {
+	gateway, err := c.getActiveGateway()
+	if err != nil {
+		klog.ErrorS(err, "Failed to get active Gateway")
+		return err
+	}
+
+	isMyselfGateway := gateway != nil && gateway.Name == c.nodeConfig.Name
+	isGatewayItem := isGatewayKey(key)
+	if isGatewayItem {
+		// Controller will try to remove the WireGuard interface if it exists on the host
+		// before WireGuard initialization or when Gateway is changed or deleted.
+		if c.wireGuardClient != nil && (!isMyselfGateway || !c.wireGuardInitialized) {
+			if err := c.cleanUpWireGuard(); err != nil {
+				return err
+			}
+			c.wireGuardInitialized = false
+		}
+		if !isMyselfGateway {
+			klog.V(4).InfoS("Current Node is not Gateway Node, skip WireGuard reconciliation")
+			return nil
+		}
+		if !c.wireGuardInitialized {
+			var initErr error
+			defer func() {
+				if initErr != nil {
+					klog.InfoS("Failed to initialize WireGuard, deleting interface and route")
+					if deleteErr := c.cleanUpWireGuard(); deleteErr != nil {
+						klog.ErrorS(deleteErr, "Failed to clean up WireGuard stale resources")
+					}
+				}
+			}()
+			if initErr = c.initializeWireGuard(); initErr != nil {
+				klog.ErrorS(initErr, "Failed to initialize WireGuard interface")
+				return initErr
+			}
+
+			if initErr = c.updateWireGuardPeers(); initErr != nil {
+				klog.ErrorS(initErr, "Failed to update WireGuard peers")
+				return initErr
+			}
+			c.wireGuardInitialized = true
+		}
+	} else {
+		if !isMyselfGateway {
+			klog.V(4).InfoS("Gateway does not exist or this Node is not Gateway, skip reconciling WireGuard")
+			return nil
+		}
+
+		// If the WireGuard has not been initialized. Some necessary configuration may not
+		// be ready. In this situation, we need to requeue the ClusterInfoImport again.
+		if !c.wireGuardInitialized {
+			return fmt.Errorf("WireGuard has not been initialized, requeue ClusterInfoImport")
+		}
+
+		ciImport, err := c.ciImportLister.ClusterInfoImports(c.namespace).Get(key)
+		if err != nil {
+			return err
+		}
+		return c.updateSingleWireGuardPeer(ciImport)
+	}
+
+	return nil
+}
+
+// updateWireGuardPeers lists all existing ClusterInfoImports and tries to add them as WireGuard peer.
+func (c *MCDefaultRouteController) updateWireGuardPeers() error {
+	klog.InfoS("Adding existing ClusterInfoImport as WireGuard peers")
+	ciImports, err := c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
```

Comment:
You should be able to use `c.ciImportLister.List(labels.Everything())` after re-base the latest codes of PR4407